cli/cmd/packageInfo.go

```diff
@@ -0,0 +1,123 @@
+		SilenceUsage: true,
+	}
+
+	return cmd
+}
+
+func decodeSigStruct(path string) error {
+	// Check if given filename is actually a directory
+	stat, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+
+	// If it is, we try to find out if it's an Occlum image directory
+	var elfFile *elf.File
+	var isOcclumInstance bool
+	if isDirectory := stat.IsDir(); isDirectory {
+		if elfFile, err = elf.Open(filepath.Join(path, "build/lib/libocclum-libos.signed.so")); err == nil {
+			color.Green("Detected Occlum image.")
+			isOcclumInstance = true
+		} else if os.IsNotExist(err) {
+			color.Red("ERROR: A directory was supplied, but it appears not to be an Occlum instance.")
+			color.Red("Please either specify the SGX enclave binary directly, or the root of an Occlum instance.")
+			return err
+		}
+	} else {
+		elfFile, err = elf.Open(path)
+	}
+	if err != nil {
+		return err
+	}
+	defer elfFile.Close()
+
+	// SIGSTRUCT is stored in ELF section '.note.sgxmeta'
+	sgxMetaSection := elfFile.Section(".note.sgxmeta")
+	if sgxMetaSection == nil {
+		return errors.New("could not find SGX metadata section (.note.sgxmeta) in given file")
+	}
+
+	sgxMetaData, err := sgxMetaSection.Data()
+	if err != nil {
+		return err
+	}
+
+	/*
+	 * From the "Intel(r) 64 and IA-32 Architectures Software Developer
+	 * Manual, Volume 3: System Programming Guide", Chapter 38, Section 13,
+	 * Table 38-19 "Layout of Enclave Signature Structure (SIGSTRUCT)"
+	 */
+	sigStructHeader := []byte{0x06, 0x00, 0x00, 0x00, 0xe1, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00}
+
+	sigStructIndex := bytes.Index(sgxMetaData, sigStructHeader)
+	if sigStructIndex == -1 {
+		return errors.New("could not find SIGSTRUCT header in given file")
+	}
+
+	// The Intel Software Developer Manual specifies SIGSTRUCT entries up to 1808 bytes.
+	// We use this as a cutoff for our sigStruct slice.
+	if len(sgxMetaData) <= sigStructIndex+1808 {
+		return errors.New("SGX metadata/SIGSTRUCT appears to be too small")
+	}
+
+	sigStruct := sgxMetaData[sigStructIndex : sigStructIndex+1808]
+
+	// SIGSTRUCT has two headers. Let's check against the second one, too.
+	// We only use the second one to verify that we actually work on the correct struct.
+	sigStructHeader2 := []byte{0x01, 0x01, 0x00, 0x00, 0x60, 0x00, 0x00, 0x00, 0x60, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00}
+	sigStructHeader2Index := bytes.Index(sigStruct, sigStructHeader2)
+	if sigStructHeader2Index == -1 {
+		return errors.New("found first SIGSTRUCT header, but cannot find second one")
+	}
+
+	// Get MRENCLAVE (UniqueID), ISVPRODID (ProductID) and ISVSVN (SecurityVersion) directly from SIGSTRUCT
+	// Get Modulus so we can calculate MRSIGNER (= SHA256 hash of modulus)
+	modulus := sigStruct[128:512]
+	mrenclave := sigStruct[960:992]
+	isvprodid := sigStruct[1024:1026]
+	isvsvn := sigStruct[1026:1028]
+
+	// Calculate MRSIGNER, which is the SHA-256 hash of the modulus stored in SIGSTRUCT
+	mrsigner := sha256.Sum256(modulus)
```

Comment:
Can you move this block in a function and add a unit test for it?

Reply:
Think this is similar to the situation we had with EGo where a test ELF file would be good or needed, however shipping one might bloat the unit tests a little bit.

If hardcoding / supplying a precompiled valid section is fine, that's no problem.

Reply:
Yes, just add a valid section as hardcoded []byte slice to the test.